processor/removeemptyvaluesprocessor/processor.go

```diff
@@ -0,0 +1,266 @@
+				span := spans.At(k)
+				cleanMap(span.Attributes(), evp.c, evp.excludeAttributeKeys)
+			}
+		}
+	}
+
+	return td, nil
+}
+
+func (evp *emptyValueProcessor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	resourceLogs := ld.ResourceLogs()
+	for i := 0; i < resourceLogs.Len(); i++ {
+		resourceLog := resourceLogs.At(i)
+		scopeLogs := resourceLog.ScopeLogs()
+
+		if evp.c.EnableResourceAttributes {
+			cleanMap(resourceLog.Resource().Attributes(), evp.c, evp.excludeResourceKeys)
+		}
+
+		for j := 0; j < scopeLogs.Len(); j++ {
+			scopeLog := scopeLogs.At(j)
+			logRecords := scopeLog.LogRecords()
+
+			for k := 0; k < logRecords.Len(); k++ {
+				logRecord := logRecords.At(k)
+				if evp.c.EnableAttributes {
+					cleanMap(logRecord.Attributes(), evp.c, evp.excludeAttributeKeys)
+				}
+
+				if evp.c.EnableLogBody {
+					cleanLogBody(logRecord, evp.c, evp.excludeBodyKeys)
+				}
+			}
+		}
+	}
+
+	return ld, nil
+}
+
+func (evp *emptyValueProcessor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	resourceMetrics := md.ResourceMetrics()
+	for i := 0; i < resourceMetrics.Len(); i++ {
+		resourceMetric := resourceMetrics.At(i)
+		scopeMetrics := resourceMetric.ScopeMetrics()
+
+		if evp.c.EnableResourceAttributes {
+			cleanMap(resourceMetric.Resource().Attributes(), evp.c, evp.excludeResourceKeys)
+		}
+
+		if !evp.c.EnableAttributes {
+			// Skip loops for attributes if we don't need to clean them.
+			continue
+		}
+
+		for j := 0; j < scopeMetrics.Len(); j++ {
+			scopeMetric := scopeMetrics.At(j)
+			metrics := scopeMetric.Metrics()
+
+			for k := 0; k < metrics.Len(); k++ {
+				metric := metrics.At(k)
+				cleanMetricAttrs(metric, evp.c, evp.excludeAttributeKeys)
+			}
+		}
+	}
+	return md, nil
+}
+
+// cleanMap removes empty values from the map, as defined by the config.
+func cleanMap(m pcommon.Map, c Config, excludeKeys []MapKey) {
+	m.RemoveIf(func(s string, v pcommon.Value) bool {
+		for _, mk := range excludeKeys {
+			if mk.key == s {
+				return false
+			}
+		}
+
+		switch v.Type() {
+		case pcommon.ValueTypeEmpty:
+			return c.RemoveNulls
+		case pcommon.ValueTypeMap:
+			subMap := v.Map()
+			cleanMap(subMap, c, trimMapKeyPrefix(s, excludeKeys))
+			return c.RemoveEmptyMaps && subMap.Len() == 0
+		case pcommon.ValueTypeSlice:
+			return c.RemoveEmptyLists && v.Slice().Len() == 0
+		case pcommon.ValueTypeStr:
+			return shouldFilterString(v.Str(), c.EmptyStringValues)
+		}
+
+		return false
+	})
+}
+
+// trimMapKeyPrefix returns the provided keys with the specified prefix removed.
+// Any keys that don't have the prefix are removed from the returned list.
+func trimMapKeyPrefix(prefix string, keys []MapKey) []MapKey {
+	outKeys := make([]MapKey, 0, len(keys))
+	for _, mk := range keys {
+		if !strings.HasPrefix(mk.key, prefix+".") {
+			// prefix was not found, so this key does not belong to the submap.
+			continue
+		}
+
+		trimmedKey := mk.key[len(prefix)+1:]
```

Comment:
nit: Might be cleaner to use `strings.TrimPrefix` here. I know it's the same thing but it's easier to read.

Reply:
I avoided TrimPrefix because it's less performant here, but maybe I can rework this block to use it (ideally I'd use CutPrefix here, but it's not available in go1.20, so I basically just inlined the code from it here)